primaryHandler.go

```diff
@@ -180,23 +286,63 @@ func NewPrimaryHandler(logger log.Logger, manager device.Manager, v *viper.Viper
 	// the secured variant of the device connect handler
 	apiHandler.Handle(
 		"/device",
-		deviceConnectChain.Append(authorizationDecoratorDevice).Then(connectHandler),
+		deviceConnectChain.
+			Extend(authChain).
+			Append(DeviceMetadataMiddleware).
+			Then(connectHandler),
 	).HeadersRegexp("Authorization", ".*")
 
-	// fail open if no authorization header is sent
 	apiHandler.Handle(
 		"/device",
-		deviceConnectChain.Then(connectHandler),
+		deviceConnectChain.
+			Append(DeviceMetadataMiddleware).
+			Then(connectHandler),
 	)
 
 	apiHandler.Handle(
 		"/device/{deviceID}/stat",
-		authorizationDecorator(&device.StatHandler{
-			Logger:   logger,
-			Registry: manager,
-			Variable: "deviceID",
-		}),
+		alice.New(
+			device.UseID.FromPath("deviceID")).
+			Extend(authChain).
+			Then(&device.StatHandler{
+				Logger:   logger,
+				Registry: manager,
+				Variable: "deviceID",
+			}),
 	)
 
 	return r, nil
 }
+
+func buildDeviceAccessCheck(config *deviceAccessCheckConfig, logger log.Logger, counter metrics.Counter, deviceRegistry device.Registry) (deviceAccess, error) {
+	errorLogger := logging.Error(logger)
+
+	if len(config.Checks) < 1 {
+		errorLogger.Log(logging.MessageKey(), "Potential security misconfig. Include a check for deviceAccessCheck or disable it")
+		return nil, errors.New("Failed enabling DeviceAccessCheck")
+	}
+
+	var parsedChecks []*parsedCheck
+	for _, check := range config.Checks {
+		parsedCheck, err := parseDeviceAccessCheck(check)
+		if err != nil {
+			errorLogger.Log(logging.ErrorKey(), err, logging.MessageKey(), "deviceAccesscheck parse failure")
+			return nil, errors.New("Failed parsing DeviceAccessChecks")
+		}
+		parsedChecks = append(parsedChecks, parsedCheck)
+	}
+
+	if config.Type == "enforce" || config.Type == "monitor" {
```

Comment:
This code could be flatter by checking at the beginning of this function to verify that the type is "enforce" or "monitor", and returning an error if it's not either.  Then, the positive case will be the one returned at the end of this function. 🙂

Reply:
Good idea, thanks!